### PR TITLE
New Cheatcode API (Moved memory management to Forge)

### LINKED ISSUE #405: Decouple cheatable starknet from forge
Cheatable starknet should be more decoupled from forge. It should expose controllers that allow interacting with starknet state but it shouldn't handle reading cheatcode inputs, writing outputs to memory, error handling etc. Instead it should just take correctly typed arguments and return `Result`s. 

Reading inputs and outputs should be handled on the Forge side.

Also, the methods of the controller shouldn't necessarily be coupled directly with specific cheatcodes and should be composable e.g. if Forge decides to introduce a new cheatcode that combines the functionality of two already existing cheatcodes, it should be possible to implement just by calling necessary cheatcodes methods.

### PR DESCRIPTION
**Pull request by Radinyn**
Closes #405
Preparation for #305

### CODE REVIEW

File: crates/forge/src/cheatcodes_hint_processor.rs [hunk 1]
```diff
@@ -209,29 +216,82 @@ impl CairoHintProcessor<'_> {
...
+                        Ok(())
+                    }
+                    Err(CheatcodeError::Recoverable(_)) => {
+                        panic!("Declare should not fail recoverably!")
+                    }
+                    Err(CheatcodeError::Unrecoverable(err)) => Err(err),
+                }
+            }
+            "deploy" => {
+                // TODO(#1991) deploy should fail if contract address provided doesn't match calculated
```
Comment (war-in) [inline crates/forge/src/cheatcodes_hint_processor.rs:262]: Are you sure about this number? 😅

**Review by war-in — APPROVED** [review]

File: crates/forge/src/cheatcodes_hint_processor.rs (hunk 1 above)
Comment (Radinyn) [reply crates/forge/src/cheatcodes_hint_processor.rs:262]: It's an old protostar issue

Comment (war-in) [reply crates/forge/src/cheatcodes_hint_processor.rs:262]: Hmm, I thought that all protostar issues were ported here

File: crates/cheatable-starknet/src/cheatcodes/declare.rs [hunk 2]
```diff
@@ -33,38 +31,40 @@ use starknet::core::types::contract::CompiledClass;
 impl CheatedState {
```
Comment (MaksymilianDemitraszek) [inline crates/cheatable-starknet/src/cheatcodes/declare.rs:31]: Would it be possible to put BlockifierState into CheatedState to abstract the separation for forge devs? `CheatedState` could be renamed to `CheatnetState`

**Comment by MaksymilianDemitraszek** [comment]
@Arcticae can you please look into that before merging?

**Review by Arcticae — APPROVED** [review]

File: crates/cheatable-starknet/src/cheatcodes/declare.rs (hunk 2 above)
Comment (Arcticae) [reply crates/cheatable-starknet/src/cheatcodes/declare.rs:31]: Seems like a valid take to me, @Radinyn WDYT?

File: crates/forge/src/cheatcodes_hint_processor.rs (hunk 1 above)
Comment (Radinyn) [reply crates/forge/src/cheatcodes_hint_processor.rs:262]: yes, but this one was closed without addressing this Todo

Comment (piotmag769) [reply crates/forge/src/cheatcodes_hint_processor.rs:262]: I think this TODO is no relevant anymore

File: crates/cheatnet/src/cheatcodes.rs [hunk 3]
```diff
@@ -45,15 +44,15 @@ impl From<EnhancedHintError> for HintError {
         }
     }
 }
+pub enum CheatcodeError {
+    Recoverable(Vec<Felt252>),        // Return result in cairo
```
Comment (piotmag769) [inline crates/cheatnet/src/cheatcodes.rs:48]:
```suggestion
    Recoverable(Vec<Felt252>), // Return error result in cairo
```

**Review by piotmag769 — APPROVED** [review]

**Pull request merged**